README.md

```diff
@@ -10,268 +10,475 @@ hard-to-find defects early in the application development lifecycle.
+      | Credentials Parameter | `CSONAR_HUBPWFILE_ID` | A **secret file** containing the password corresponding to `CSONAR_HUBUSER`. |
+
+   * Option 2: Hub user certificate (HTTPS hubs only).
+
+     If the hub user account you want to use for authenticating the build/analysis does not already have a user certificate and private key, configure one from the **User Certificates** page before proceeding.
+
+      | **Type (select from Add Parameter pulldown)** | **Name**  | **Value** |
+      |------------------|---------------------------|-------------------------|
+      | Credentials Parameter | `CSONAR_USER_CERT_ID` | A **secret file** containing a *user certificate* (PEM format) for the hub user account. This will be presented to the hub for authentication. |
+      | Credentials Parameter | `CSONAR_USER_PRIVKEY_ID` | A **secret file** containing the *private key* (PEM format) corresponding to `CSONAR_USER_CERT_ID`. This will be used to sign challenges from the hub: it will *not* be presented to the hub. |
+
+      **MANUAL:** How CodeSonar Works > CodeSonar Structure > TLS Certificates
+
+      **MANUAL:**  Using CodeSonar > GUI Reference > Hub and Account Management > GUI: User Certificates
+1. Edit the build stage of your Jenkins pipeline integrate the CodeSonar build/analysis.
+   1. Edit the agent specification to add the `CodeSonar` label (or whatever label you added to the agents where you installed CodeSonar.)
+   1. Wrap the software build command or commands in a `withCredentials()` block.
+      * Password authentication:
+        ```
+        withCredentials([file(credentialsId: params.CSONAR_HUBPWFILE_ID, variable: 'CSONAR_HUBPWFILE')]) { ... }
+        ```
+      * Certificate authentication:
+        ```
+        withCredentials([file(credentialsId: params.CSONAR_USER_CERT_ID, variable: 'CSONAR_USER_CERT'),
+                          file(credentialsId: params.CSONAR_USER_PRIVKEY_ID, variable: 'CSONAR_USER_PRIVKEY')]) { ... }
+        ```
+    1. Set up environment variable `CSONAR_CMD_ARGS` to collect together the various `codesonar` command line elements.
+       ```
+       env.CSONAR_CMD_ARGS = "$CSONAR_HUB_ADDRESS \
+                              $CSONAR_WORKDIR/$CSONAR_PROJECT_NAME \
+                              -project $CSONAR_PROJECT_NAME \
+                              -auth password -hubuser $CSONAR_HUBUSER -hubpwfile $CSONAR_HUBPWFILE \
+                              -foreground"
+       ```
+       * For certificate-based authentication, replace the `-auth` line with:
+         ```
+         -auth certificate -hubcert $CSONAR_USER_CERT -hubkey $CSONAR_USER_PRIVKEY \
+         ```
+       * For CodeSonar SaaS, replace the `-foreground` flag with:
+         ```
+         -remote /saas/*
+         ```
+       * For (non-SaaS) remote-managed analyses, replace the `-foreground` flag with:
+         ```
+         -remote <analysis-launchd>
+         ```
```

Comment:
We must also use `-wait` with `-remote`.  Otherwise, the plugin will attempt to fetch analysis results before they are ready.